### DISCUSSION TITLE
Cache cleared when error occurs

### DISCUSSION THREAD

**Discussion by debater-coder** [TanStack/query#4119] [Q&A] [answered]
Hi, I'm making an offline PWA that uses `persistQueryClient` to cache queries in `localStorage`. My issue is, that when an error occurs, the cache for that query is cleared. I would like to find a way to disable this behavior so that the user can keep working with the existing data, even if they are disrupted by a potentially temporary error.

**Comment by TkDodo** [comment, marked as answer]
> My issue is, that when an error occurs, the cache for that query is cleared.

you mean when the query errors? error are not json serializable so you cannot write them to localstorage. You can use something like `superjson` to serialize / deserialize.

**Reply by debater-coder** [reply]
My desired behavior is that the existing data stays in the cache, instead of being evicted. I don't want to write the error to cache. Is there any way to make this happen?

**Reply by TkDodo** [reply]
I don't understand why data should be evicted from the in-memory cache. If a query errors, it:
- goes into error state
- but should still show stale data if you have data from a previous fetch

however, queries in error state are _not_ persisted to localstorage. So if you reload the page then, that data is gone.

If you experience different behaviour, please show it in a codesandbox reproduction.

**Reply by debater-coder** [reply]
Is there any way to keep the data after the page is reloaded?

**Reply by debater-coder** [reply]
From the TanStack Query docs:
> [Removal](https://tanstack.com/query/v4/docs/plugins/persistQueryClient#removal)
> If data is found to be any of the following:
> - expired (see maxAge)
> - busted (see buster)
> - *error (ex: throws ...)*
> - empty (ex: undefined)
> the persister removeClient() is called and the cache is immediately discarded.

How can I change the *error* behaviour? Is there any way to do so? What is the reasoning behind this behaviour?

**Reply by TkDodo** [reply]
these docs are about what happens when storing to localstorage errors - _not_ when a query errors.

You can also dehydrate queries that are in error state, however, keep in mind that the `error` itself can likely not be serialized if it's of type `Error` unless you do custom serialization.

`shouldDehydrateQuery` defaults to:

```
function defaultShouldDehydrateQuery(query: Query) {
  return query.state.status === 'success'
}
```

you can change that by passing a custom function:

```
<PersistQueryClientProvider
  client={queryClient}
  persistOptions={{
    persister,
    dehydrateOptions: { shouldDehydrateQuery: () => true },
  }}
>
```
